ui/app/utils/supervised_fine_tuning/openAITokenCounter.ts

```diff
@@ -0,0 +1,294 @@
+// TypeScript port of the token counting example using Tiktoken from the Python version of OpenAI Cookbook
+// Reference: https://cookbook.openai.com/examples/how_to_count_tokens_with_tiktoken
+
+import { encoding_for_model, get_encoding } from "tiktoken";
+import type { TiktokenModel } from "tiktoken";
+import { MODEL_TOKEN_LIMITS, CURRENT_MODEL_VERSIONS } from "./constants";
+import type { OpenAIMessage, ToolFunction } from "./types";
+
+/**
+ * Converts model name to tiktoken model name
+ * @param model Model name to convert
+ * @returns Corresponding tiktoken model name
+ */
+export function convertToTiktokenModel(
+  model: string,
+): TiktokenModel | undefined {
+  // Check if the model is already a valid TiktokenModel
+  try {
+    // Try to create an encoder with the model name
+    const encoder = encoding_for_model(model as TiktokenModel);
+    encoder.free();
+    return model as TiktokenModel;
+  } catch {
+    // If the model is not valid, proceed with model family checks
+  }
+
+  // Handle model families
+  if (model.startsWith("gpt-4o-")) {
+    return "gpt-4";
+  }
+  if (model.startsWith("gpt-4-")) {
+    return "gpt-4";
+  }
+  if (model.startsWith("gpt-3.5-turbo")) {
+    return "gpt-3.5-turbo";
+  }
+
+  return;
+}
+
+/**
+ * Gets token limit for a given model
+ * @param model Model name
+ * @returns Token limit for the model
+ */
+export function getModelTokenLimit(model: string): number {
+  // Check exact model match first
+  if (model in MODEL_TOKEN_LIMITS) {
+    return MODEL_TOKEN_LIMITS[model];
+  }
+
+  // Handle model families
+  if (model.startsWith("gpt-4o-")) {
+    return MODEL_TOKEN_LIMITS["gpt-4o-2024-08-06"];
+  }
+  if (model.startsWith("gpt-4-")) {
+    return MODEL_TOKEN_LIMITS["gpt-4-0613"];
+  }
+  if (model.startsWith("gpt-3.5-turbo")) {
+    return MODEL_TOKEN_LIMITS["gpt-3.5-turbo-0125"];
+  }
+
+  // Default to conservative limit for unknown models
+  return 4096;
+}
+
+/**
+ * Gets encoding for a given model, falling back to o200k_base for unknown models
+ * @param model Model name
+ * @returns Tiktoken encoding
+ */
+function getEncodingForModel(model: string) {
+  const tiktokenModel = convertToTiktokenModel(model);
+  if (!tiktokenModel) {
+    console.warn(`Unknown model: ${model}, using o200k_base`);
+    return get_encoding("o200k_base");
+  }
+  return encoding_for_model(tiktokenModel);
+}
+
+/**
+ * Calculate the number of tokens used by a list of messages
+ * @param messages Array of messages to count tokens for
+ * @param model Model name to use for token counting
+ * @returns Number of tokens used by the messages
+ */
+export function getTokensFromMessages(
```

Comment:
can you cite where this calculation is taken from (I believe the Python codebase has a citation in it)